modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkByScrollAction.java

```diff
@@ -354,7 +354,7 @@ private void recordFailure(Failure failure, List<Failure> failures) {
      * Start terminating a request that finished non-catastrophically.
      */
     void startNormalTermination(List<Failure> indexingFailures, List<ShardSearchFailure> searchFailures, boolean timedOut) {
-        if (task.isCancelled() || false == mainRequest.isRefresh()) {
+        if (task.isCancelled() || false == mainRequest.isRefresh() || destinationIndices.isEmpty()) {
```

Comment:
If you give it a query that filters out all the docs. Or you use a script to convert all the changes to noops.

`destinationIndices` is updated based on the generated IndexRequests so it can handle sneaky things like scripts changing the destination index. I suspect it is just a bonus that we can skip the refresh on noops.